### ISSUE TITLE
Why API should be NSObject and why framework takes responsibility for maintaining singleton?

### ISSUE THREAD

**Issue by garnett** [ishkawa/APIKit#13]
IMO, singleton implementation based on class name is not obvious and tricky (it forces API to be NSObject subclass) + sometimes users of this kinds of libs do not want use singletons (for example for testing).

Also, the type of ```Github.instance``` in demo app is not ```Github``` but ```API```, that's a bug.

**Comment by ishkawa** [comment]
- Why API should be NSObject?
-> To be a delegate of NSURLSession.

- Why framework takes responsibility for maintaining singleton?
-> I intended to provide separated sessions for each subclasses of APIs.

Actually, this implementation is tricky and not obvious, so changing this implementation is acceptable for me. How about this idea?

Managing Instance:
- API does not have any instances and session dictionary.
- API has a default session as a static variable.
- If an user want to separate session for a subclass, this static variable should be overrided.
  (This means the user takes responsibility for maintaining instance.)

Default implementation of NSURLSessionDelegate:
- The default session has a delegate (not API instance) that implements [these methods](https://github.com/ishkawa/APIKit/blob/0601c7389df19c02d1d58c1cca82b12e3d6478c7/APIKit/APIKit.swift#L196-L208).
- If an user want to add custom behavior using delegate methods in NSURLSessionDelegate, session variable should be overrided by custom session with a delegate that implements additional methods.

Pros:
- No weird implementation.

Cons:
- More steps to implement custom delegate methods.

BTW, description of `GitHub.instance` returns `<DemoApp.GitHub: 0x7fb34272e9f0>` at runtime, but it is not inferred as `GitHub` while coding. Anyway, this will be disappear from APIKit if the change above is adopted.

**Comment by garnett** [comment]
I've forgot completely that NSURLSession delegate should be NSObject 😇 
In general described solution looks much more obvious for me, but I still have concerns about this case:
let's say I'm working with github API, and by project requirements _some_ requests should utilize background session and all other - default. How can I solve that without duplicating code in my sibilings of API? The only way I see is make hieararchy ```API <- Githib <- BackgroundGithub```

**Comment by ishkawa** [comment]
Good point. The case you mentioned should be considered in APIKit. Associating NSURLSession with API on 1-1 relation might make this library less flexible. On the other hand, APIKit should have 1 session at least for the simplest usage. Based on them, how about this:

- API has a default session, and `sendRequest(request:handler:)` method uses this session.
- To send request in the other session, call `sendRequest(request:URLSession:handler:)`.

```swift
public class API {
    public var defaultURLSession: NSURLSession
    public class func sendRequest(request:handler:)
    public class func sendRequest(request:URLSession:handler:)
}
```

**Comment by ishkawa** [comment]
I have just pushed https://github.com/ishkawa/APIKit/commit/ad0b537c2db70cc8e841881dd48b6590235e2b75 to feature branch, which implements this idea.

**Comment by ishkawa** [comment]
This is resolved in 0.4.0.
